DeviceCode/Targets/Native/STM32F4/DeviceCode/Initialization/tinyhal.cpp

```diff
@@ -0,0 +1,752 @@
+
+                ::Watchdog_ResetCounter();
+
+                // write back to sector, as we only change one bit from 0 to 1, no need to erase sector
+                bRet = (TRUE == pBlockDevice->Write( configSectAddress, pBlockRegionInfo->BytesPerBlock, data, FALSE ));
+
+                private_free(data);
+            }
+        }
+
+        CPU_Reset();
+    }
+}
+
+bool g_fDoNotUninitializeDebuggerPort = false;
+
+void HAL_Initialize()
+{    
+#if defined(PLATFORM_ARM_OS_PORT)
+    // Interrupts must be enabled to handle calls to OS
+    // (Network stack uses the CMSIS-RTX OS, which uses
+    // SVC calls, which will hard fault if the interrupts
+    // are disabled at the Svc instruction )
+    // SystemInit handles this for the startup from reset
+    // However, this is also called from the CLR when doing
+    // a soft reboot.
+    __enable_irq();
+#endif
+	
+    HAL_CONTINUATION::InitializeList();
+    HAL_COMPLETION  ::InitializeList();
+
+    HAL_Init_Custom_Heap();
+
+    Time_Initialize();
+    Events_Initialize();
+
+    CPU_GPIO_Initialize();
+    CPU_SPI_Initialize();
+
+#if !defined(PLATFORM_ARM_OS_PORT)
+    // this is the place where interrupts are enabled after boot for the first time after boot
+    ENABLE_INTERRUPTS();
+#endif
+
+    // have to initialize the blockstorage first, as the USB device needs to update the configure block
+
+    BlockStorageList::Initialize();
+
+    BlockStorage_AddDevices();
+
+    BlockStorageList::InitializeDevices();
+
+    //FS_Initialize();
+
+    //FileSystemVolumeList::Initialize();
+
+    //FS_AddVolumes();
+
+    //FileSystemVolumeList::InitializeVolumes();
+
+    //LCD_Initialize();
+    
+#if !defined(HAL_REDUCESIZE)
+    CPU_InitializeCommunication();
+#endif
+
+    I2C_Initialize();
+
+    Buttons_Initialize();
+
+    // Initialize the backlight to a default off state
+    //BackLight_Initialize();
+
+    //Piezo_Initialize();
+
+    //Battery_Initialize();
+
+    //Charger_Initialize();
+
+    PalEvent_Initialize();
+    //Gesture_Initialize();
+    //Ink_Initialize();
+    TimeService_Initialize();
+
+#if defined(ENABLE_NATIVE_PROFILER)
+    Native_Profiler_Init();
+#endif
+}
+
+void HAL_UnReserveAllGpios()
+{
+    for(INT32 i = CPU_GPIO_GetPinCount()-1; i >=0; i--)
+    {
+        CPU_GPIO_ReservePin((GPIO_PIN)i, false);
+    }
+}
+
+#if defined(PLATFORM_ARM_OS_PORT) && defined(TCPIP_LWIP_OS)
+// Hack: refer to work item #2374
+// For reasons unknown, the ARM linker is getting the
+// fixup for the LoadRegion symbol incorrect. The data
+// is located correctly but the fixed up pointer stored
+// in the literal pool that this code loads for the address
+// of the load base (src) is off by some factor. In initial
+// testing it was always 0x90, unfortunately it turns out
+// not to be consistent and bumped up to 0xED0, and is now
+// back at 0x90... Sigh... Hope to hear back from ARM support
+// on this soon.
+const UINT32 ArmLinkerLoadRegionOffsetHack = 0x00000090;

```

Comment:
This hack can't go here. It is completely dependent on the entire set of code and data for the full firmware image, and therefore VERY solution specific.